### ISSUE TITLE
function for nothing / error data information in master

### ISSUE THREAD

**Issue by petrovr** [openssl/openssl#9756]
Before as part of error code was encoded "function code".
This functionality was removed, error related macros replaced with new one...

Nasty side effect is that ERR_error_string* stop to return function!
So far so good, this does not impact my code - it extracts "error data".

The defect is that "error data" does not return information for function, i.e. with current code base information for function is complelly lost.

Build with no-err. As part of this function and reason tables are not loaded. But now functionality is based on compiler feature. So fix of this issue may change this. Function name is not private information.

In addition could be updated macros ERR_PACK and ERR_GET_FUNC do not encode/decode "function code".
Function ERR_func_error_string could could be changed return empty string and marked as deprecated. Return values "system library" or NULL are not relevant.

Macro OPENSSL_FUNC:
* ```PEDANTIC``` is internal define and cannot be used in public header.
* ```__FUNCSIG__``` - hmm.. All those macros should defined only in function body. I guess that check fail outside function. "pretty function" name should be visible when is visible ```__FUNCTION__```.

Off topic...
It seems  to me current development model allows a feature(functionality) to be removed without implementation of replacement functionality. This model is even worse than deprecation.

P.S. sample result from an program with OpenSSL current master (2 Sep 2019):
```
 eng_try_load_cert: crypto message: error:8000007A:E_NSS::Unsupported key type:KEYTYPE=116
[ OpenSSL error message start here  ^, old function here^,                    ^ custom data - separator (:)  and data from error]
```

**Comment by petrovr** [comment]
Also mkerr.pl script is mostly useless - it does not include function codes into generated table. Result is that this script cannot be used at all to write engine that supports earlier versions.
The code generated with script from previous release with small modifications could be adapted to for 3.0. 
From updated  script is expected to generate code that is backward compatible.

**Comment by levitte** [comment]
> Before as part of error code was encoded "function code".
> This functionality was removed, error related macros replaced with new one...

Yup.  It was discussed quite a bit on openssl-project: https://mta.openssl.org/pipermail/openssl-project/2019-June/001426.html
It was also eventually voted upon: https://mta.openssl.org/pipermail/openssl-project/2019-July/001465.html

The idea is that the error record contains two sets of information:

1.  Error data, which is error location (library, unit, what-you-wanna-call-it) code and reason code and extra error information, which are fairly normal bits of information one might want programmatically or to somehow display what error it was.  Those bits are all that's needed to get a full error string back to the application.
2.  Debugging information, which is the file name, the line number and where available, the function name.

So what has happened is that we've traded function codes for function names in string form.  Function codes were programmatically useless, since they could change at any time because of code refactoring.

*Something I realise, though, is that while we're saving the function name string, we're never giving it back to the application*, i.e. the changes are incomplete.
Thanks, this issue was instrumental for this realisation.

**Comment by levitte** [comment]
> Also mkerr.pl script is mostly useless

Most of all, it's really an internal script.  External software have always had to create their own solutions...  or take a copy of mkerr.pl and do the best they can.
Yes, having written a few engines for external customers a number of years ago, I know *exactly* how painful that is.

> Result is that this script cannot be used at all to write engine that supports earlier versions.

We aren't claiming downward compatibility, so have done nothing to mitigate for different OpenSSL versions.  Considering it's an internal script, it is of course only adapted to the matching OpenSSL version an none other.

**Comment by petrovr** [comment]
I understand that "function codes" prevent monster refactoring like "move to provider".

The point here is that before [FUNC] always is part of error. It is not debug information. 
When program request string representation of message for [FUNC] it will contain either a) function code or b) function name.

Case b) is normal build and program ask library to load strings. Sample:
```
error:02001002:system library:fopen:No such file or directory @@@ fopen('/file/not/found','r')
error:2006D080:BIO routines:BIO_new_file:no such file
```
Remark @@@ separates data

Case a) is when program does not load string or openssl is build with no-err. Then The above messages look like:
```
error:02001002:lib(2):func(1):reason(2) @@@ fopen('/file/not/found','r')
error:2006D080:lib(32):func(109):reason(128)
```

Switch to [FUNC] provided by compiler does not mean that information for "current function name" has to be ignored in "no-err" build, i.e. method ERR_set_debug shall be called with function argument.


Of course main point of issue is that current code lack functionality to extract [FUNC].
As side note is to think about improvement in development(deprecation) model - first implement then clean-up obsolete.

**Comment by levitte** [comment]
I have a PR for you to look at: #9765
